File: frontend/src/utils/artistEventSupply.js
/**
 * Pick the single event that should display the "most tickets" badge.
 * When multiple events tie for max supply, the earliest upcoming date wins.
 *
 * @param {Array<{ id: string|number, tickets_count?: number, date?: string }>} events
 * @returns {string|number|null}
 */
export function pickMostSupplyEventId(events) {
  if (!Array.isArray(events) || events.length === 0) return null;

  const max = Math.max(...events.map((ev) => Number(ev?.tickets_count) || 0));
  if (max <= 0) return null;

  const tied = events.filter((ev) => (Number(ev?.tickets_count) || 0) === max);
  if (tied.length === 1) return tied[0].id ?? null;

  const earliest = [...tied].sort((a, b) => new Date(a.date) - new Date(b.date))[0];
  return earliest?.id ?? null;
}

File: backend/users/tests/_artist_event_supply.py
"""Mirror of frontend pickMostSupplyEventId — keeps badge tie-break spec testable in Django."""


def pick_most_supply_event_id(events):
    if not events:
        return None
    max_count = max(int(ev.get('tickets_count') or 0) for ev in events)
    if max_count <= 0:
        return None
    tied = [ev for ev in events if int(ev.get('tickets_count') or 0) == max_count]
    if len(tied) == 1:
        return tied[0].get('id')
    earliest = sorted(tied, key=lambda ev: ev.get('date') or '')[0]
    return earliest.get('id')
